qiskit_experiments/library/characterization/multi_state_discrimination.py

```diff
@@ -0,0 +1,121 @@
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Multi state discrimination experiment."""
+
+from typing import Dict, List, Optional
+
+from qiskit import QuantumCircuit
+from qiskit.circuit import Gate
+from qiskit.providers import Backend
+from qiskit.providers.options import Options
+from qiskit.pulse import ScheduleBlock
+from qiskit.qobj.utils import MeasLevel
+from qiskit_experiments.framework import BaseExperiment
+from qiskit_experiments.library.characterization import MultiStateDiscriminationAnalysis
+
+
+class MultiStateDiscrimination(BaseExperiment):
+    r"""An experiment that prepares the first :math:`n+1` energy states,
+    i.e. :math:`|0\rangle,\cdots,|n\rangle`. TODO:is the name ok? not discriminating yet.
+
+    # section: example
+
+        The steps to run a multi state discrimination experiment are TODO
+
+    # section: reference
+        `Qiskit Textbook <https://qiskit.org/textbook/ch-quantum-hardware/accessing\
+        _higher_energy_states.html>`_.
+
+    # section: tutorial
+        :doc:`TODO?`
+    """
+
+    @classmethod
+    def _default_run_options(cls) -> Options:
+        """Default option values for the experiment :meth:`run` method."""
+        options = super()._default_run_options()
+
+        options.meas_level = MeasLevel.KERNELED
+        options.meas_return = "single"
+
+        return options
+
+    @classmethod
+    def _default_experiment_options(cls) -> Options:
+        """Default values for the number of states if none is given.
+
+        Experiment Options:
+            n_states (int): The number of excited states.
+
+        """
+        options = super()._default_experiment_options()
+        options.n_states = 2
+
+        return options
+
+    def __init__(
+            self,
+            qubit: int,
+            backend: Optional[Backend] = None,
+            n_states: Optional[int] = None,
+            schedules: Optional[Dict[str, ScheduleBlock]] = None
+    ):
+        """Setup an experiment to prepare different energy states on a given qubit.
+
+        Args:
+            qubit: The qubit on which to run the experiment.
+            backend: Optional, the backend to run the experiment on.
+            n_states: The number of energy levels to prepare.
+            schedules: The schedules to reach each energy level.
+        """
+
+        super().__init__((qubit,), analysis=MultiStateDiscriminationAnalysis(), backend=backend)
+
+        self.experiment_options.schedules = schedules
+
+        if n_states is None:
+            n_states = self.experiment_options.n_states
+        self.set_experiment_options(n_states=n_states)
```

Comment:
```suggestion
        if n_states is not None:
            self.set_experiment_options(n_states=n_states)
```

Reply:
Addressed in commit [9390381](https://github.com/Qiskit/qiskit-experiments/pull/936/commits/9390381ce44f62e19b8243b76bd212b37df0fa8f).